src/operator/mshadow_op.h

```diff
@@ -1068,12 +574,12 @@ struct smooth_l1_loss {
   template<typename DType>
   MSHADOW_XINLINE static DType Map(DType a, DType b) {
     b *= b;
-    if (a > DType(1.0f) / b) {
-      return a - DType(0.5f) / b;
-    } else if (a < DType(-1.0f) / b) {
-      return -a - DType(0.5f) / b;
+    if (a > 1.0f / b) {
```

Comment:
This is in the kernel; calculating `1.0f/b` twice is expensive. Maybe caching the result of `1.0f/b` can save some time.

Reply:
Hello,

all code below gammaln_grad is old code (before any of my changes). In fact, I rolled this code back, because somehow my previous changes gave somebody trouble.

Because of this, I'd like to leave all the code below gammaln_grad as is (it is not mine, anyway). But please do feel free to improve this code in another PR. Just make sure it does not break anything elsewhere ...

Reply:
I think the compiler is smart enough to optimize this out

Reply:
This smooth_l1 code is very odd, not covered by a single test